File: NeuronMCP/internal/database/query.go
package database

import (
	"fmt"
	"strings"
)

// QueryBuilder provides utilities for building SQL queries
type QueryBuilder struct{}

// Select builds a SELECT query
func (qb *QueryBuilder) Select(table string, columns []string, where map[string]interface{}, orderBy *OrderBy, limit, offset *int) (string, []interface{}) {
	if len(columns) == 0 {
		columns = []string{"*"}
	}

	var params []interface{}
	paramIndex := 1

	// SELECT clause
	selectClause := strings.Join(columns, ", ")

	// FROM clause
	fromClause := EscapeIdentifier(table)

	// WHERE clause
	var whereClause string
	if len(where) > 0 {
		var conditions []string
		for key, value := range where {
			escapedKey := EscapeIdentifier(key)
			conditions = append(conditions, fmt.Sprintf("%s = $%d", escapedKey, paramIndex))
			params = append(params, value)
			paramIndex++
		}
		whereClause = "WHERE " + strings.Join(conditions, " AND ")
	}

	// ORDER BY clause
	var orderByClause string
	if orderBy != nil {
		orderByClause = fmt.Sprintf("ORDER BY %s %s", EscapeIdentifier(orderBy.Column), orderBy.Direction)
	}

	// LIMIT clause
	var limitClause string
	if limit != nil {
		limitClause = fmt.Sprintf("LIMIT $%d", paramIndex)
		params = append(params, *limit)
		paramIndex++
	}

	// OFFSET clause
	var offsetClause string
	if offset != nil {
		offsetClause = fmt.Sprintf("OFFSET $%d", paramIndex)
		params = append(params, *offset)
	}

	parts := []string{
		"SELECT " + selectClause,
		"FROM " + fromClause,
		whereClause,
		orderByClause,
		limitClause,
		offsetClause,
	}

	var nonEmptyParts []string
	for _, part := range parts {
		if part != "" {
			nonEmptyParts = append(nonEmptyParts, part)
		}
	}

	query := strings.Join(nonEmptyParts, " ")
	return query, params
}

// OrderBy represents an ORDER BY clause
type OrderBy struct {
	Column    string
	Direction string // ASC or DESC
}

// VectorSearch builds a vector search query
func (qb *QueryBuilder) VectorSearch(table, vectorColumn string, queryVector []float32, distanceMetric string, limit int, additionalColumns []string, minkowskiP *float64) (string, []interface{}) {
	var params []interface{}
	paramIndex := 1

	// Convert vector to string format
	vectorStr := formatVector(queryVector)
	params = append(params, vectorStr)

	var operator string
	var distanceExpr string

	switch distanceMetric {
	case "cosine":
		operator = "<=>"
		distanceExpr = fmt.Sprintf("%s %s $%d::vector AS distance", EscapeIdentifier(vectorColumn), operator, paramIndex)
	case "inner_product":
		operator = "<#>"
		distanceExpr = fmt.Sprintf("%s %s $%d::vector AS distance", EscapeIdentifier(vectorColumn), operator, paramIndex)
	case "l1":
		distanceExpr = fmt.Sprintf("vector_l1_distance(%s, $%d::vector) AS distance", EscapeIdentifier(vectorColumn), paramIndex)
	case "hamming":
		distanceExpr = fmt.Sprintf("vector_hamming_distance(%s, $%d::vector) AS distance", EscapeIdentifier(vectorColumn), paramIndex)
	case "chebyshev":
		distanceExpr = fmt.Sprintf("vector_chebyshev_distance(%s, $%d::vector) AS distance", EscapeIdentifier(vectorColumn), paramIndex)
	case "minkowski":
		p := 2.0
		if minkowskiP != nil {
			p = *minkowskiP
		}
		paramIndex++
		params = append(params, p)
		distanceExpr = fmt.Sprintf("vector_minkowski_distance(%s, $%d::vector, $%d::double precision) AS distance", EscapeIdentifier(vectorColumn), paramIndex-1, paramIndex)
	default: // l2
		operator = "<->"
		distanceExpr = fmt.Sprintf("%s %s $%d::vector AS distance", EscapeIdentifier(vectorColumn), operator, paramIndex)
	}

	paramIndex++
	selectColumns := []string{"*"}
	for _, col := range additionalColumns {
		selectColumns = append(selectColumns, EscapeIdentifier(col))
	}

	params = append(params, limit)
	limitClause := fmt.Sprintf("LIMIT $%d", paramIndex)

	query := fmt.Sprintf(`
		SELECT %s, %s
		FROM %s
		ORDER BY distance
		%s
	`, strings.Join(selectColumns, ", "), distanceExpr, EscapeIdentifier(table), limitClause)

	return query, params
}

// formatVector formats a float32 slice as a PostgreSQL vector string
func formatVector(vec []float32) string {
	var parts []string
	for _, v := range vec {
		parts = append(parts, fmt.Sprintf("%g", v))
	}
	return "[" + strings.Join(parts, ",") + "]"
}

File: NeuronMCP/src/database/query.ts
/**
 * Query builder utilities for safe SQL construction
 */

export class QueryBuilder {
	/**
	 * Build SELECT query
	 */
	static select(
		table: string,
		columns: string[] = ["*"],
		where?: Record<string, any>,
		orderBy?: { column: string; direction?: "ASC" | "DESC" },
		limit?: number,
		offset?: number
	): { query: string; params: any[] } {
		const params: any[] = [];
		let paramIndex = 1;

		// SELECT clause
		const selectClause = columns.map((col) => this.escapeIdentifier(col)).join(", ");

		// FROM clause
		const fromClause = this.escapeIdentifier(table);

		// WHERE clause
		let whereClause = "";
		if (where && Object.keys(where).length > 0) {
			const conditions = Object.entries(where).map(([key, value]) => {
				const escapedKey = this.escapeIdentifier(key);
				if (Array.isArray(value)) {
					return `${escapedKey} = ANY($${paramIndex++})`;
				} else {
					return `${escapedKey} = $${paramIndex++}`;
				}
			});
			whereClause = `WHERE ${conditions.join(" AND ")}`;
			params.push(...Object.values(where));
		}

		// ORDER BY clause
		let orderByClause = "";
		if (orderBy) {
			orderByClause = `ORDER BY ${this.escapeIdentifier(orderBy.column)} ${orderBy.direction || "ASC"}`;
		}

		// LIMIT clause
		let limitClause = "";
		if (limit !== undefined) {
			limitClause = `LIMIT $${paramIndex++}`;
			params.push(limit);
		}

		// OFFSET clause
		let offsetClause = "";
		if (offset !== undefined) {
			offsetClause = `OFFSET $${paramIndex++}`;
			params.push(offset);
		}

		const query = [
			`SELECT ${selectClause}`,
			`FROM ${fromClause}`,
			whereClause,
			orderByClause,
			limitClause,
			offsetClause,
		]
			.filter(Boolean)
			.join(" ");

		return { query, params };
	}

	/**
	 * Build INSERT query
	 */
	static insert(table: string, data: Record<string, any>): { query: string; params: any[] } {
		const columns = Object.keys(data);
		const values = Object.values(data);
		const params: any[] = [];

		const columnList = columns.map((col) => this.escapeIdentifier(col)).join(", ");
		const valuePlaceholders = values.map((_, i) => `$${i + 1}`).join(", ");
		params.push(...values);

		const query = `INSERT INTO ${this.escapeIdentifier(table)} (${columnList}) VALUES (${valuePlaceholders}) RETURNING *`;

		return { query, params };
	}

	/**
	 * Build UPDATE query
	 */
	static update(
		table: string,
		data: Record<string, any>,
		where: Record<string, any>
	): { query: string; params: any[] } {
		const params: any[] = [];
		let paramIndex = 1;

		const setClause = Object.entries(data)
			.map(([key]) => {
				return `${this.escapeIdentifier(key)} = $${paramIndex++}`;
			})
			.join(", ");
		params.push(...Object.values(data));

		const whereConditions = Object.entries(where)
			.map(([key]) => {
				return `${this.escapeIdentifier(key)} = $${paramIndex++}`;
			})
			.join(" AND ");
		params.push(...Object.values(where));

		const query = `UPDATE ${this.escapeIdentifier(table)} SET ${setClause} WHERE ${whereConditions} RETURNING *`;

		return { query, params };
	}

	/**
	 * Build DELETE query
	 */
	static delete(table: string, where: Record<string, any>): { query: string; params: any[] } {
		const params: any[] = [];
		let paramIndex = 1;

		const whereConditions = Object.entries(where)
			.map(([key]) => {
				return `${this.escapeIdentifier(key)} = $${paramIndex++}`;
			})
			.join(" AND ");
		params.push(...Object.values(where));

		const query = `DELETE FROM ${this.escapeIdentifier(table)} WHERE ${whereConditions} RETURNING *`;

		return { query, params };
	}

	/**
	 * Escape SQL identifier
	 */
	static escapeIdentifier(identifier: string): string {
		return `"${identifier.replace(/"/g, '""')}"`;
	}

	/**
	 * Build vector search query
	 */
	static vectorSearch(
		table: string,
		vectorColumn: string,
		queryVector: number[],
		distanceMetric: "l2" | "cosine" | "inner_product" | "l1" | "hamming" | "chebyshev" | "minkowski" = "l2",
		limit: number = 10,
		additionalColumns: string[] = [],
		minkowskiP?: number
	): { query: string; params: any[] } {
		const vectorStr = `[${queryVector.join(",")}]`;
		let operator = "<->";
		let distanceFunction = "vector_l2_distance";
		const params: any[] = [vectorStr];

		switch (distanceMetric) {
			case "cosine":
				operator = "<=>";
				distanceFunction = "vector_cosine_distance";
				break;
			case "inner_product":
				operator = "<#>";
				distanceFunction = "vector_inner_product";
				break;
			case "l1":
				distanceFunction = "vector_l1_distance";
				break;
			case "hamming":
				distanceFunction = "vector_hamming_distance";
				break;
			case "chebyshev":
				distanceFunction = "vector_chebyshev_distance";
				break;
			case "minkowski":
				distanceFunction = "vector_minkowski_distance";
				if (minkowskiP === undefined) {
					minkowskiP = 2;
				}
				params.push(minkowskiP);
				break;
		}

		const selectColumns = ["*", ...additionalColumns.map((col) => this.escapeIdentifier(col))];
		
		// For operators, use operator syntax; for functions, use function call
		let distanceExpr: string;
		if (distanceMetric === "minkowski") {
			distanceExpr = `${distanceFunction}(${this.escapeIdentifier(vectorColumn)}, $1::vector, $2::double precision) AS distance`;
		} else if (["l1", "hamming", "chebyshev"].includes(distanceMetric)) {
			distanceExpr = `${distanceFunction}(${this.escapeIdentifier(vectorColumn)}, $1::vector) AS distance`;
		} else {
			distanceExpr = `${this.escapeIdentifier(vectorColumn)} ${operator} $1::vector AS distance`;
		}

		params.push(limit);

		const query = `
			SELECT ${selectColumns.join(", ")}, ${distanceExpr}
			FROM ${this.escapeIdentifier(table)}
			ORDER BY distance
			LIMIT $${params.length}
		`;

		return { query, params };
	}
}

